### ISSUE TITLE
Dictionary<string,byte[]> formatter is being trimmed with PublishAot

### ISSUE THREAD

**Issue by dusrdev** [Cysharp/MemoryPack#272]
Seems that the formatter that is being used internally in MemoryPack for `Serialize(Dictionary<string,byte[])` is being trimmed, and the use of such object is now entirely broken.

<img width="1134" alt="Screenshot 2024-03-28 at 16 03 06" src="https://github.com/Cysharp/MemoryPack/assets/8972626/8311ac8c-5f61-4a4d-96c6-595c2b150010">

**Comment by dusrdev** [comment]
Is there any update?

I have even attempted to somehow extract and create an instance of the required formatter myself, as I know the types at design time, but it has proven to be very difficult since a lot of the api's require use of inaccessible internal MemoryPack types/properties...

I'd even settle for a workaround, specifically for `Dictionary<string, byte[]>`, without this my entire app is broken...

**Comment by meryuhi** [comment]
Maybe you can put following code somewhere to prevent trimming.
```csharp
MemoryPackFormatterProvider.RegisterDictionary<Dictionary<string, byte[]>, string, byte[]>();
```
Or define a `MemoryPackable(GenerateType.Collection)` type that extends from `Dictionary<string, byte[]>`.

**Comment by dusrdev** [comment]
> Maybe you can put following code somewhere to prevent trimming.
> 
> ```cs
> MemoryPackFormatterProvider.RegisterDictionary<Dictionary<string, byte[]>, string, byte[]>();
> ```
> 
> Or use a `MemoryPackable(GenerateType.Collection)` type that extends from `Dictionary<string, byte[]>`.

The formatters for IMemoryPackable<T> implementing types are also being trimmed. Is looks like it is possible to register the formatters like you noted, but with many types, it seems verbose, and should probably happen automatically, this is the entire idea behind the attributes I assume...

**Comment by meryuhi** [comment]
Yes, it should happen automatically as you said, registering the formatters manually is just a workaround.
I guess it could be solved by #251...
